src/middlewared/middlewared/plugins/mail.py

```diff
@@ -1,41 +1,324 @@
 from middlewared.schema import Bool, Dict, Int, List, Str, accepts
-from middlewared.service import ConfigService
+from middlewared.service import CallError, ConfigService, ValidationErrors, job, periodic, private
 
+from datetime import datetime, timedelta
+from email.message import Message
+from email.mime.multipart import MIMEMultipart
+from email.mime.text import MIMEText
+from email.utils import formatdate
+from lockfile import LockFile, LockTimeout
+
+import base64
+import errno
+import json
 import os
-import sys
+import pickle
+import smtplib
+import socket
+import syslog
+
+
+class QueueItem(object):
+
+    def __init__(self, message):
+        self.attempts = 0
+        self.message = message
+
+
+class MailQueue(object):
+
+    QUEUE_FILE = '/tmp/mail.queue'
+    MAX_ATTEMPTS = 3
+
+    def __init__(self):
+        self.queue = None
+
+    def append(self, message):
+        self.queue.append(QueueItem(message))
+
+    @classmethod
+    def is_empty(cls):
+        if not os.path.exists(cls.QUEUE_FILE):
+            return True
+        try:
+            return os.stat(cls.QUEUE_FILE).st_size == 0
+        except OSError:
+            return True
+
+    def _get_queue(self):
+        try:
+            with open(self.QUEUE_FILE, 'rb') as f:
+                self.queue = pickle.loads(f.read())
+        except (pickle.PickleError, EOFError):
+            self.queue = []
 
-if '/usr/local/www' not in sys.path:
-    sys.path.append('/usr/local/www')
-os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'freenasUI.settings')
+    def __enter__(self):
+        self._lock = LockFile(self.QUEUE_FILE)
+        while not self._lock.i_am_locking():
+            try:
+                self._lock.acquire(timeout=330)
+            except LockTimeout:
+                self._lock.break_lock()
 
-import django
-from django.apps import apps
-if not apps.ready:
-    django.setup()
+        if not os.path.exists(self.QUEUE_FILE):
+            open(self.QUEUE_FILE, 'a').close()
 
-from freenasUI.common.system import send_mail
+        self._get_queue()
+        return self
+
+    def __exit__(self, typ, value, traceback):
+
+        with open(self.QUEUE_FILE, 'wb+') as f:
+            if self.queue:
+                f.write(pickle.dumps(self.queue))
```

Comment:
We should probably write to `f{self.QUEUE_FILE}.tmp` and then rename it, otherwise queue may break on power/system failure

Reply:
Why? QUEUE_FILE is in /tmp which is a memory file system recreated on every boot.

Reply:
@william-gr makes sense, my mistake